app/src/behaviors/behavior_smart_interrupt.c

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright (c) 2022 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_smart_interrupt
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+#include <zmk/keymap.h>
+#include <zmk/matrix.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/position_state_changed.h>
+#include <zmk/events/layer_state_changed.h>
+#include <zmk/hid.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#define ZMK_BHV_MAX_ACTIVE_SMART_INTERRUPTS 10
+
+struct behavior_smart_interrupt_config {
+    int32_t shared_key_positions_len;
+    int32_t shared_layers_len;
+    struct zmk_behavior_binding *behaviors;
+    int32_t shared_layers[32];
+    int32_t timeout_ms;
+    int32_t shared_key_positions[];
```

Comment:
Ditto, could be a bitmask instead of this "sparse" array.

Reply:
From the discussion on discord this is a bit tricky since the keymap can be any length. I've changed it to 8-bit ints for now but will keep working on a better solution.